primitives/transaction/src/account/htlc_contract.rs

```diff
@@ -269,52 +323,219 @@ impl OutgoingHTLCTransactionProof {
+                        Ok(AnyHash::Sha512(hash))
+                    }
+                    _ => Err(A::Error::invalid_value(
+                        serde::de::Unexpected::Str(algorithm),
+                        &"an AnyHash variant",
+                    )),
+                }
+            } else {
+                let algorithm: u8 = seq
+                    .next_element()?
+                    .ok_or_else(|| A::Error::invalid_length(0, &self))?;
+                match algorithm {
+                    1u8 => {
+                        let hash: AnyHash32 = seq
+                            .next_element()?
+                            .ok_or_else(|| A::Error::invalid_length(1, &self))?;
+                        Ok(AnyHash::Blake2b(hash))
+                    }
+                    3u8 => {
+                        let hash: AnyHash32 = seq
+                            .next_element()?
+                            .ok_or_else(|| A::Error::invalid_length(1, &self))?;
+                        Ok(AnyHash::Sha256(hash))
+                    }
+                    4u8 => {
+                        let hash: AnyHash64 = seq
+                            .next_element()?
+                            .ok_or_else(|| A::Error::invalid_length(1, &self))?;
+                        Ok(AnyHash::Sha512(hash))
+                    }
+                    _ => Err(A::Error::invalid_value(
+                        serde::de::Unexpected::Unsigned(algorithm as u64),
+                        &"an AnyHash variant",
+                    )),
+                }
+            }
+        }
+    }
+
+    impl<'de> Deserialize<'de> for AnyHash {
         fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
         where
             D: Deserializer<'de>,
         {
-            let data: Cow<'de, str> = Deserialize::deserialize(deserializer)?;
-            data.parse().map_err(Error::custom)
+            let human_readable = deserializer.is_human_readable();
+            deserializer.deserialize_struct(
+                "AnyHash",
+                ANYHASH_FIELDS,
+                AnyHashVisitor { human_readable },
+            )
         }
     }
 
-    impl Serialize for AnyHash64 {
+    impl Serialize for PreImage {
         fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
         where
             S: Serializer,
         {
-            serializer.serialize_str(&self.to_hex())
+            if serializer.is_human_readable() {
+                match self {
+                    PreImage::PreImage32(hash) => Serialize::serialize(hash, serializer),
+                    PreImage::PreImage64(hash) => Serialize::serialize(hash, serializer),
+                }
+            } else {
+                let mut state = serializer.serialize_struct("PreImage", PREIMAGE_FIELDS.len())?;
+                match self {
+                    PreImage::PreImage32(pre_image) => {
+                        state.serialize_field(PREIMAGE_FIELDS[0], &0u8)?;
```

Comment:
I believe it's usual for the byte to express the length of the following field. So this number should be `&32u8`, and for `PreImage64` it should be `&64u8`. I think that makes more sense than `0` and `1`.

Reply:
Done @sisou